src/commands/signzone.rs

```diff
@@ -0,0 +1,285 @@
+        }
+
+        //---
+
+        let (apex, ttl) = Self::find_apex(&records).unwrap();
+
+        let opt_out = if self.hash_only {
+            Nsec3OptOut::OptOut
+        } else if self.nsec3_opt_out_flags_only {
+            Nsec3OptOut::OptOutFlagsOnly
+        } else {
+            Nsec3OptOut::NoOptOut
+        };
+
+        let hashes = if self.use_nsec3 {
+            let params = Nsec3param::new(self.algorithm, 0, self.iterations, self.salt.clone());
+            let Nsec3Records {
+                recs,
+                param,
+                hashes,
+            } = records
+                .nsec3s::<_, BytesMut>(&apex, ttl, params, opt_out, self.diagnostic_comments)
+                .unwrap();
+            records.extend(recs.into_iter().map(Record::from_record));
+            records.insert(Record::from_record(param)).unwrap();
+            hashes
+        } else {
+            let nsecs = records.nsecs::<Bytes>(&apex, ttl);
+            records.extend(nsecs.into_iter().map(Record::from_record));
+            None
+        };
+
+        if !self.hash_only {
+            let inception: Timestamp = Timestamp::now().into_int().sub(10).into();
+            let expiration = inception.into_int().add(2592000).into(); // XXX 30 days
+            let extra_records = records
+                .sign(&apex, expiration, inception, keys.as_slice())
+                .unwrap();
+            records.extend(extra_records.into_iter().map(Record::from_record));
+        }
+
+        if let Some(hashes) = hashes {
+            records
+                .write_with_comments(&mut std::io::stdout().lock(), |r| {
+                    if r.rtype() == Rtype::NSEC3 {
+                        hashes.get(r.owner()).map(|hash| format!(" {hash}"))
+                    } else {
+                        None
+                    }
+                })
+                .unwrap();
+        } else {
+            records.write(&mut std::io::stdout().lock()).unwrap();
+        }
+
+        Ok(())
+    }
+
+    fn load_zone(&self) -> Result<SortedRecords<StoredName, StoredRecordData>, Error> {
+        let mut zone_file = File::open(self.zonefile_path.as_path())?;
+        let mut reader = inplace::Zonefile::load(&mut zone_file).unwrap();
+        if let Some(origin) = &self.origin {
+            reader.set_origin(origin.to_owned());
+        }
+        let mut records = SortedRecords::new();
+        for entry in reader {
+            let Ok(entry) = entry else {
+                return Err(Error::from(format!(
+                    "Invalid zone file: {}",
+                    entry.unwrap_err()
+                )));
+            };
+            match entry {
+                Entry::Record(record) => {
+                    let record: StoredRecord = record.flatten_into();
+                    records
+                        .insert(record)
+                        .expect("Invalid zone file: Duplicate record detected: {record:?}");
```

Comment:
An error that the user could cause shouldn't result in a panic.  Besides, I don't think the format string here will actually be interpolated.  Prefer `.map_err` and the `?` operator.

Reply:
Fair point.